File: game/sdk/python/src/sdk/logger.py
"""The logger APIs.

The logger API is the API that can be used to log messages.
"""

import sys
import threading
from abc import ABC, abstractmethod
from datetime import datetime

from termcolor import cprint


class ILogger(ABC):
    """The logger interface.

    The logger interface is the interface that can be used to log messages.
    """
    @abstractmethod
    def debug(self, message: str) -> None:
        """Logs a debug message.

        Args:
            message (str): The message to log.
        """
        raise NotImplementedError

    @abstractmethod
    def info(self, message: str) -> None:
        """Logs an info message.

        Args:
            message (str): The message to log.
        """
        raise NotImplementedError

    @abstractmethod
    def warn(self, message: str) -> None:
        """Logs a warning message.

        Args:
            message (str): The message to log.
        """
        raise NotImplementedError

    @abstractmethod
    def error(self, message: str) -> None:
        """Logs an error message.

        Args:
            message (str): The message to log.
        """
        raise NotImplementedError


class Logger(ILogger):
    _FORMAT = ""

    _lock = threading.Lock()

    def __init__(self, namespace: str) -> None:
        super().__init__()

        gettrace = getattr(sys, "gettrace", None)
        if gettrace is not None and gettrace():
            self._debug = True
        else:
            self._debug = False

        self._namespace = namespace

    def debug(self, message: str) -> None:
        if not self._debug:
            return
        
        with Logger._lock:
            cprint(f'{Logger._get_current_time_string()} ', color='cyan', end='')
            cprint(f'DEBUG ', color='dark_grey', end='')
            cprint(f'[{self._namespace}] {message}', color='dark_grey', end='')
            print()

    def info(self, message: str) -> None:
        with Logger._lock:
            cprint(f'{Logger._get_current_time_string()} ', color='cyan', end='')
            cprint(f'INFO  ', color='blue', end='')
            cprint(f'[{self._namespace}] {message}', color='white', end='')
            print()

    def warn(self, message: str) -> None:
        with Logger._lock:
            cprint(f'{Logger._get_current_time_string()} ', color='cyan', end='')
            cprint(f'WARN  ', color='yellow', end='')
            cprint(f'[{self._namespace}] {message}', color='yellow', end='')
            print()

    def error(self, message: str) -> None:
        with Logger._lock:
            cprint(f'{Logger._get_current_time_string()} ', color='cyan', end='')
            cprint(f'ERROR ', color='red', end='')
            cprint(f'[{self._namespace}] {message}', color='red', end='')
            print()

    @staticmethod
    def _get_current_time_string() -> str:
        return datetime.now().strftime(f'%H:%M:%S')

File: server/src/GameServer/Utilities/Logger/Logger.cs
using System.Diagnostics;

namespace GameServer.Utilities.Logger;

/// <summary>
/// Logger class provides logging functionality.
/// </summary>
public class Logger : ILogger
{
    #region Fields and properties
    private readonly string _loggingNamespace;
    #endregion

    #region Constructors and finalizers
    /// <summary>
    /// Creates a new Logger instance.
    /// </summary>
    public Logger(string loggingNamespace)
    {
        _loggingNamespace = loggingNamespace;
    }
    #endregion

    #region Methods
    /// <summary>
    /// Logs an debug message.
    /// </summary>
    /// <param name="message">The message</param>
    public void Debug(string message)
    {
        lock (Console.Out)
        {
            PrintDebug(message);
        }
    }

    /// <summary>
    /// Logs an information message.
    /// </summary>
    /// <param name="message">The message</param>
    public void Info(string message)
    {
        lock (Console.Out)
        {
            Print($"{GetCurrentTimeString()} ", ConsoleColor.Cyan);
            Print($"INFO  ", ConsoleColor.Blue);
            Print($"[{_loggingNamespace}] {message}", ConsoleColor.White);
            Console.WriteLine();
        }
    }

    /// <summary>
    /// Logs an warning message.
    /// </summary>
    /// <param name="message">The message</param>
    public void Warning(string message)
    {
        lock (Console.Out)
        {
            Print($"{GetCurrentTimeString()} ", ConsoleColor.Cyan);
            Print($"WARN  ", ConsoleColor.Yellow);
            Print($"[{_loggingNamespace}] {message}", ConsoleColor.Yellow);
            Console.WriteLine();
        }
    }

    /// <summary>
    /// Logs an error message.
    /// </summary>
    /// <param name="message">The message</param>
    public void Error(string message)
    {
        lock (Console.Out)
        {
            Print($"{GetCurrentTimeString()} ", ConsoleColor.Cyan);
            Print($"ERROR ", ConsoleColor.Red);
            Print($"[{_loggingNamespace}] {message}", ConsoleColor.Red);
            Console.WriteLine();
        }
    }

    [Conditional("DEBUG")]
    private void PrintDebug(string message)
    {
        Print($"{GetCurrentTimeString()} ", ConsoleColor.Cyan);
        Print($"DEBUG ", ConsoleColor.Gray);
        Print($"[{_loggingNamespace}] {message}", ConsoleColor.Gray);
        Console.WriteLine();
    }

    /// <summary>
    /// Gets the current time string.
    /// </summary>
    /// <returns>The current time string</returns>
    private static string GetCurrentTimeString()
    {
        return DateTime.Now.ToString("HH:mm:ss");
    }

    /// <summary>
    /// Prints a message.
    /// </summary>
    /// <param name="message">The message</param>
    /// <param name="messageColor">The message color</param>
    private static void Print(string message, ConsoleColor messageColor)
    {
        Console.ForegroundColor = messageColor;
        Console.Write(message);
        Console.ResetColor();
    }
    #endregion
}
